providers/implementations/ciphers/cipher_sm4_gcm_hw.c

```diff
@@ -14,37 +14,46 @@
 #include "cipher_sm4_gcm.h"
 #include "crypto/sm4_platform.h"
 
+# define SM4_GCM_HW_SET_KEY_CTR_FN(ks, fn_set_enc_key, fn_block, fn_ctr)       \
+    ctx->ks = ks;                                                              \
+    fn_set_enc_key(key, ks);                                                   \
+    CRYPTO_gcm128_init(&ctx->gcm, ks, (block128_f)fn_block);                   \
+    ctx->ctr = (ctr128_f)fn_ctr;                                               \
+    ctx->key_set = 1;
+
 static int sm4_gcm_initkey(PROV_GCM_CTX *ctx, const unsigned char *key,
                            size_t keylen)
 {
     PROV_SM4_GCM_CTX *actx = (PROV_SM4_GCM_CTX *)ctx;
     SM4_KEY *ks = &actx->ks.ks;
 
-    ctx->ks = ks;
 # ifdef HWSM4_CAPABLE
     if (HWSM4_CAPABLE) {
-        HWSM4_set_encrypt_key(key, ks);
-        CRYPTO_gcm128_init(&ctx->gcm, ks, (block128_f) HWSM4_encrypt);
 #  ifdef HWSM4_ctr32_encrypt_blocks
-        ctx->ctr = (ctr128_f) HWSM4_ctr32_encrypt_blocks;
+        SM4_GCM_HW_SET_KEY_CTR_FN(ks, HWSM4_set_encrypt_key, HWSM4_encrypt,
+                                  HWSM4_ctr32_encrypt_blocks);
 #  else /* HWSM4_ctr32_encrypt_blocks */
-        ctx->ctr = (ctr128_f)NULL;
+        SM4_GCM_HW_SET_KEY_CTR_FN(ks, HWSM4_set_encrypt_key, HWSM4_encrypt, NULL);
 #  endif
     } else
 # endif /* HWSM4_CAPABLE */
+
+#ifdef VPSM4_EX_CAPABLE
+    if (VPSM4_EX_CAPABLE) {
+        SM4_GCM_HW_SET_KEY_CTR_FN(ks, vpsm4_ex_set_decrypt_key, vpsm4_ex_encrypt,
```

Comment:
@zzl360 sm4-gcm test failures when I run `make TESTS=test_evp test` on Kunpeng-920, maybe the `fn_set_enc_key` should be `vpsm4_ex_set_encrypt_key` here?

Reply:
yes, I'll fix it soon.

Reply:
> yes, I'll fix it soon.

#20751